core/src/main/java/org/apache/iceberg/mapping/MappingUtil.java

```diff
@@ -114,18 +122,45 @@ private MappedFields addNewFields(MappedFields mapping, int parentId) {
         return mapping;
       }
 
-      List<MappedField> fields = Lists.newArrayList();
-      if (mapping != null) {
-        fields.addAll(mapping.fields());
-      }
-
+      List<MappedField> newFields = Lists.newArrayList();
       for (Types.NestedField add : fieldsToAdd) {
         MappedFields nestedMapping = TypeUtil.visit(add.type(), CreateMapping.INSTANCE);
-        fields.add(MappedField.of(add.fieldId(), add.name(), nestedMapping));
+        newFields.add(MappedField.of(add.fieldId(), add.name(), nestedMapping));
+      }
+
+      if (mapping == null || mapping.fields().isEmpty()) {
+        return MappedFields.of(newFields);
+      }
+
+      ImmutableMap.Builder<String, Integer> builder = ImmutableMap.builder();
+      fieldsToAdd.stream().forEach(field -> builder.put(field.name(), field.fieldId()));
+      Map<String, Integer> assignments = builder.build();
+
+      // create a copy of fields that can be updated (append new fields, replace existing for reassignment)
+      List<MappedField> fields = Lists.newArrayList();
+      for (MappedField field : mapping.fields()) {
+        fields.add(removeReassignedNames(field, assignments));
       }
 
+      fields.addAll(newFields);
+
       return MappedFields.of(fields);
     }
+
+    private static MappedField removeReassignedNames(MappedField field, Map<String, Integer> assignments) {
+      MappedField newField = field;
+      for (String name : field.names()) {
+        Integer assignedId = assignments.get(name);
+        if (assignedId != null && !Objects.equals(assignedId, field.id())) {
+          newField = removeName(field, name);
+        }
+      }
+      return newField;
+    }
+
+    private static MappedField removeName(MappedField field, String name) {
```

Comment:
does it make sense to drop a MappedField if the `Set.difference` is empty?

Reply:
`testDeleteAndAddColumnReassign` highlights this problem where  mappedField can end up being empty

```
[ {
  "field-id" : 1,
  "names" : [ ]
}, {
  "field-id" : 2,
  "names" : [ "data" ]
}, {
  "field-id" : 3,
  "names" : [ "id" ]
} ]
```

Reply:
Unlike the case where we have names without an ID, I think we _could_ remove these. But I'm reluctant to because it throws away information -- that the field is known to the mapping, but doesn't have any names. That can be used to detect that a field was replaced if you wanted to. I don't have a concrete use case for it, but I think I'm leaning toward keeping it in case.

Reply:
Sounds good.